proposals/compile-time-dependency-resolution.md

```diff
@@ -0,0 +1,489 @@
+# Compile time dependency resolution
+
+* **Type**: Design proposal
+* **Author**: Raul Raja
+* **Contributors**: Tomás Ruiz López, Jorge Castillo, Francesco Vasco, Claire Neveu
+* **Status**: New
+* **Prototype**: [initial implementation](https://github.com/arrow-kt/kotlin/pull/6)
+
+## Summary
+
+The goal of this proposal is to enable **compile time dependency resolution** through extension syntax. Overall, we'd want to enable extension contract interfaces to be defined as function or class constructor arguments and enable compiler to automatically resolve and inject those instances that must be provided evidence for in one of a given set of scopes. In case of not having evidence of any of those required interfaces (program constraints), compiler would fail and provide proper error messages.
+
+## Motivation
+
+* Support compile-time verification of program dependencies (extensions).
+* Enable nested extension resolution.
+* Support compile-time verification of a program correctness given behavioral constraints are raised to the interface types.
+* Enable definition of polymorphic functions whose constraints can be verified at compile time in call sites.
+
+## Description
+
+We propose to use the existing `interface` semantics, allowing for generic definition of behaviors and their instances in the same style interfaces are defined.
+
+```kotlin
+package com.data
+
+interface Repository<A> {
+  fun loadAll(): List<A>
+  fun loadById(id: Int): A?
```

Comment:
we should add an example with `fun A.save(): Unit` to show this is actually about grouping extensions too.

Reply:
Done in fba1db3 along with another change suggested by @pakoito 👍